### PR TITLE
Build notebook dependency graph for `%run` cells

### LINKED ISSUE #1201: [FEATURE]: Detect notebook include graph by analysing `%run` magic cells
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

From https://docs.databricks.com/en/notebooks/notebook-workflows.html:

The %run command allows you to include another notebook within a notebook. You can use %run to modularize your code, for example by putting supporting functions in a separate notebook. You can also use it to concatenate notebooks that implement the steps in an analysis. When you use %run, the called notebook is immediately executed and the functions and variables defined in it become available in the calling notebook.

image

### Proposed Solution

Detect `%run` magic cells and treat them as a dependency

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
 - create a Notebook class that parses source code and splits it into cells
 - create a NotebookDependencyGraph class and related utils
 - discover dependencies in `%run` cells

### Linked issues
Resolves #1201 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
+from __future__ import annotations  # for type hints
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:1]: do we really need it?..

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        try:
+            ast = parse_python(self._original_code)
+            return ast is not None
+        except SyntaxError:
+            return False
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        pass  # TODO
+        # see https://github.com/databrickslabs/ucx/issues/1200
+        # see https://github.com/databrickslabs/ucx/issues/1202
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:57]:
```suggestion
        # TODO https://github.com/databrickslabs/ucx/issues/1200
        # TODO https://github.com/databrickslabs/ucx/issues/1202
        pass
```

small nit - i'm planning to create issues from TODO automatically and let's have them follow the same'ish convention

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+            if start >= 0:
+                path = line[start + len(command) :].strip()
+                parent.register_dependency(path.strip('"'))
+                return
+        raise ValueError("Missing notebook path in %run command")
+
+
+class CellLanguage(Enum):
+    # long magic_names must come first to avoid shorter ones being matched
+    PYTHON = Language.PYTHON, 'python', '#', PythonCell
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:139]: oh, we can do that?..

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                line = line[len(magic_language_prefix) :]
+                return CellLanguage.of_magic_name(line.strip())
+            return None
+        return None
+
+    def new_cell(self, source: str) -> Cell:
+        return self._new_cell(source)
+
+
+def extract_cells(source: str, default_language: CellLanguage) -> list[Cell] | None:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:193]: shouldn't it be a method on `CellLanguage`?

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    cell_lines: list[str] = []
+    separator = f"{default_language.comment_prefix}{CELL_SEPARATOR}"
+    for i in range(1, len(lines)):
+        line = lines[i].strip()
+        if line.startswith(separator):
+            cell = make_cell(cell_lines)
+            cells.append(cell)
+            cell_lines = []
+        else:
+            cell_lines.append(lines[i])
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:222]:
```suggestion
            continue
        cell_lines.append(lines[i])
```

File: tests/unit/source_code/test_files.py [hunk 6]
```diff
@@ -68,7 +68,7 @@ def test_files_walks_directory():
     languages.linter(Language.PYTHON).lint.return_value = [Mock(code='some-code')]
     languages.fixer.return_value = None
     files = Files(languages)
-    path = Path(__file__).parent
+    path = Path(Path(__file__).parent.parent, "aws/")
```
Comment (ericvergnaud) [inline tests/unit/source_code/test_files.py:71]: seems unrelated but fails with this PR because source_code folder now contains non-python files

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        return self._new_cell(source)
+
+
+def extract_cells(source: str, default_language: CellLanguage) -> list[Cell] | None:
+    lines = source.split('\n')
+    header = f"{default_language.comment_prefix}{NOTEBOOK_HEADER}"
+    if not lines[0].startswith(header):
+        raise ValueError("Not a Databricks notebook source!")
+
+    def make_cell(lines_: list[str]):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:199]: probably better as a private method on CellLanguage

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+
+    @classmethod
+    def of_language(cls, language: Language) -> CellLanguage:
+        return next((cl for cl in CellLanguage if cl.language == language))
+
+    @classmethod
+    def of_magic_name(cls, magic_name: str) -> CellLanguage | None:
+        return next((cl for cl in CellLanguage if magic_name.startswith(cl.magic_name)), None)
+
+    def read_cell_language(self, lines: list[str]) -> CellLanguage | None:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:173]:
```suggestion
    def of_lines(self, lines: list[str]) -> CellLanguage | None:
```

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+        self._path = path
+        self._parent = parent
+        self._locator = locator
+        self._dependencies: dict[str, NotebookDependencyGraph] = {}
+
+    @property
+    def path(self):
+        return self._path
+
+    def register_dependency(self, path: str) -> NotebookDependencyGraph:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:242]:
```suggestion
    def register(self, path: str) -> NotebookDependencyGraph:
```
`_dependency` is a bit redundant

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+            self._dependencies[path] = child_graph
+            return child_graph
+        # nay, create the child graph and populate it
+        child_graph = NotebookDependencyGraph(path, self, self._locator)
+        self._dependencies[path] = child_graph
+        notebook = self._locator(path)
+        notebook.build_dependency_graph(child_graph)
+        return child_graph
+
+    def locate_dependency(self, path: str) -> NotebookDependencyGraph | None:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:255]:
```suggestion
    def locate(self, path: str) -> NotebookDependencyGraph | None:
```

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+        else:
+            cell_lines.append(lines[i])
+    if len(cell_lines) > 0:
+        cell = make_cell(cell_lines)
+        cells.append(cell)
+
+    return cells
+
+
+class NotebookDependencyGraph:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:230]:
```suggestion
class DependencyGraph:
```
might be less redundant this way - we're in `notebooks.py` anyway

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+MAGIC_PREFIX = ' MAGIC'
+LANGUAGE_PREFIX = ' %'
+
+
+class Cell(ABC):
+
+    def __init__(self, source: str):
+        self._original_code = source
+
+    @property
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:23]:
```suggestion
```
code migration should be an explicit action, properties are implicit actions and preferably immutable

File: tests/unit/source_code/notebooks/01_HL7Streaming.scala [hunk 13]
```diff
@@ -0,0 +1,221 @@
+// Databricks notebook source
```
Comment (nfx) [inline tests/unit/source_code/notebooks/01_HL7Streaming.scala:1]: let's be consistent and add `.sample` suffix also to non-python notebooks. 

you can also add `.txt` :)

File: tests/unit/source_code/test_files.py (hunk 6 above)
Comment (nfx) [reply tests/unit/source_code/test_files.py:71]: doesn't seem to be relevant.

File: tests/unit/source_code/test_notebook.py [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+import pytest
+from databricks.sdk.service.workspace import Language
+
+from databricks.labs.ucx.source_code.notebook import Notebook, NotebookDependencyGraph
+from tests.unit import _load_sources
+
+PYTHON_NOTEBOOK_SAMPLE = (
+    "00_var_context.py.sample",
+    Language.PYTHON,
+    ['md', 'md', 'md', 'python', 'python', 'python', 'md', 'python', 'md'],
```
Comment (nfx) [inline tests/unit/source_code/test_notebook.py:12]: can you remove the irrelevant test cells?

File: tests/unit/source_code/test_notebook.py [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+    ],
+)
+R_NOTEBOOK_SAMPLE = (
+    "3_SparkR_Fine Grained Demand Forecasting.r",
+    Language.R,
+    [
+        'md',
+        'md',
+        'md',
+        'r',
```
Comment (nfx) [inline tests/unit/source_code/test_notebook.py:85]: looks too long and noisy as for the test sample

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:139]: you mean class enums ?
yes we can ! 
(I might win the next US elections with this catch phrase)

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:193]: I don't think so, it doesn't change from 1 language to another, except for the comment prefix. CellLanguage is really about those variations. But it wouldn't be bad.

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:222]: that's really a preference, correct ? Isn't it easier to reason about if/elif/else ?

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 12 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:23]: that would return the migrated_code once it's generated. I'm not planning to implement its generation here. This property is needed for checking consolidation of the notebook from small bits.

File: tests/unit/source_code/test_notebook.py (hunk 14 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_notebook.py:12]: mmm... it was actually super useful to use real notebooks to discover edge cases, and testing against an untouched sample creates a lot of confidence

File: tests/unit/source_code/test_notebook.py (hunk 15 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_notebook.py:85]: maybe we can move these to json files ?

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:1]: yes for forward type hints

File: tests/unit/source_code/test_files.py (hunk 6 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_files.py:71]: see my comment, this PR breaks the test

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:222]: I also don't like https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/no-else-return.html

File: tests/unit/source_code/notebooks/01_HL7Streaming.scala (hunk 13 above)
Comment (ericvergnaud) [reply tests/unit/source_code/notebooks/01_HL7Streaming.scala]: done

File: tests/unit/source_code/test_notebook.py (hunk 14 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_notebook.py:12]: it'd be much less intrusive if `make fmt` didn't force 1 item per line. Is there a way to configure this ?

File: tests/unit/source_code/test_notebook.py (hunk 15 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_notebook.py]: I'm not alone, see https://github.com/psf/black/issues/1281
Fixed by disabling `fmt` for these lines

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py]: Second thoughts, I think you're right. I changed it.

File: tests/unit/source_code/test_notebook.py (hunk 14 above)
Comment (ericvergnaud) [reply tests/unit/source_code/test_notebook.py:13]: Fixed by disabling fmt for these lines

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:1]: without it, you have to use `'MyType'` instead of `MyType`, and `'MyType' or None` instead of `MyType | None`

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 8 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py:173]: not sure about this one, the `of_xxx` methods are static and selecting an Enum value from their property, this instance method has a very different contract

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 11 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py]: Yeah I was hesitant about this when writing it... done

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 9 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py]: true if you're looking at it from the DependencyGraph class, not so much if you're looking at it from a caller point of view:  
`parent.register_dependency(path.strip('"'))`
is much better imho than
`parent.register(path.strip('"'))`

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 10 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebook.py]: same here

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 3 above)
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebook.py:139]: didn't realise we can just pass a tuple in there

File: src/databricks/labs/ucx/source_code/notebook.py (hunk 8 above)
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebook.py:173]: this method is still static, no?

**Review by nfx — APPROVED** [review]
let's merge this
